### PR TITLE
add a secret ref for webhook secrets

### PR DESCRIPTION
**Pull request by bparees**
bug 1504819

https://bugzilla.redhat.com/show_bug.cgi?id=1504819

### CODE REVIEW

**Comment by bparees** [comment]
@deads2k ptal/merge.  needed for https://github.com/openshift/origin/pull/17314

File: build/v1/types.go [hunk 1]
```diff
@@ -899,14 +899,27 @@ type BuildConfigStatus struct {
 	LastVersion int64 `json:"lastVersion" protobuf:"varint,1,opt,name=lastVersion"`
 }
 
+// SecretLocalReference contains information that points to the local secret being used
+type SecretLocalReference struct {
+	// Name is the name of the resource being referenced
```
Comment (deads2k) [inline build/v1/types.go]: "name of the secret in the same namespace being reference"

**Comment by deads2k** [comment]
Lifting my comments from the earlier thread:

General thoughts:
 1. This prescribes the key, but doesn't add a type.  Adding types (but not requiring them) can help users categorize their secrets.  That could happen as a later step.
 2. Being prescriptive to reduce choice simplifies the API, but as you noted it can make discoverability harder.  If you created a `secretKey` which defaulted to `WebHookSecretKey` would be a backwards compatible change that allows people to avoid specifying if they wish, but may help discoverability.  This could happen later if we actually do have problems.  I personally like being prescriptive.
 3. It is now possible to have a webhook configuration which is invalid at runtime.  With other resources, we use controllers to reflect this in status.  A similar approach could be taken here or you could write-back status from the rest handler.

This looks like a good starting point to me.  We can grow on this in a few directions if we end up needing to.

@bparees any thoughts on how to manage the status aspect?  A small controller seems like an option.  

@smarterclayton how do you generally feel about controllers watching the validity of references over time and updating status in general?  It seems consistent with other approaches and reasonably discoverable.  It would simplify status reporting the CLI and the webconsole too.

**Comment by deads2k** [comment]
> @deads2k ptal/merge. needed for openshift/origin#17314

Travis seems stuck.  I'll see if it unsticks or if we must summon @stevekuznetsov and get some real CI.

**Comment by bparees** [comment]
> @bparees any thoughts on how to manage the status aspect? A small controller seems like an option.

we already have secret refs (And imagestream refs) elsewhere in the buildconfig that aren't validated.  A general buildconfig controller that validated all the refs in the buildconfig and provided status(or events?) isn't a bad idea, but won't be part of this specific refactor.

File: build/v1/types.go (hunk 1 above)
Comment (bparees) [reply build/v1/types.go]: fixed.

**Comment by bparees** [comment]
seems to have passed travis after my update.

**Comment by bparees** [comment]
@deads2k anything else?

**Comment by bparees** [comment]
(I suppose expecting we have bots setup for this repo would be too much)

**Comment by deads2k** [comment]
> we already have secret refs (And imagestream refs) elsewhere in the buildconfig that aren't validated. A general buildconfig controller that validated all the refs in the buildconfig and provided status(or events?) isn't a bad idea, but won't be part of this specific refactor.

I agree.  Just wanted to know about general general concept.

/lgtm

**Comment by deads2k** [comment]
> (I suppose expecting we have bots setup for this repo would be too much)

Welcome to the wild west.  I'll fix that :)

**Comment by stevekuznetsov** [comment]
Yeah there is no way to give the bots access to all repos, and they need write access to a repo to do anything meaningful to it, so just ask me or someone else who can edit those configs to add them.

**Pull request merged**